pkg/controller/ingressstate/ingress_state_controller.go

```diff
@@ -0,0 +1,247 @@
+	kubeInformersForTargetNamespace.Core().V1().Endpoints().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *IngressStateController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(ingressStateControllerWorkQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(ingressStateControllerWorkQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(ingressStateControllerWorkQueueKey) },
+	}
+}
+
+// checkPodStatus will check the target pod container status and return list of possible problems.
+func (c *IngressStateController) checkPodStatus(reference *corev1.ObjectReference) []string {
+	pod, err := c.podGetter.Pods(reference.Namespace).Get(reference.Name, metav1.GetOptions{})
+	if err != nil {
+		return []string{fmt.Sprintf("error getting pod %q: %v", reference.Name, err)}
+	}
+	result := []string{}
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		if containerStatus.Ready && containerStatus.State.Running != nil {
+			continue
+		}
+		if terminated := containerStatus.State.Terminated; terminated != nil {
+			result = append(result, fmt.Sprintf("pod %q container %q terminated with %q", pod.Name, containerStatus.Name, terminated.Message))
+		}
+		if containerStatus.RestartCount > 1 {
+			result = append(result, fmt.Sprintf("pod %q container %q restarted %d times", pod.Name, containerStatus.Name, containerStatus.RestartCount))
+		}
+		// TODO: Add more here, like we can detect "pending" pods
+	}
+	return result
+}
+
+// checkEndpointHealthz will check the health of given https://endpointIP:6443/healthz
+// TODO: this is insecure for now, but enough to tell us if we can get through
+func (c *IngressStateController) checkEndpointHealthz(endpointIP string) error {
+	// TODO: consider making this secure, however it will mean wiring the trust which must match the same trust as ingress controller
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	response, err := http.Get(fmt.Sprintf("https://%s:6443/healthz", endpointIP))
+	if err != nil {
+		reportedError := err
+		if response != nil {
+			reportedError = fmt.Errorf("status:%q, body: %q, error: %v", response.Status, response.Body, reportedError)
+		}
+		return reportedError
+	}
+	if response.StatusCode != http.StatusOK {
+		return fmt.Errorf("status:%q, body: %q", response.Status, response.Body)
+	}
+	return nil
+}
+
+// degradedConditionNames lists all conditions that this controller manage
+var degradedConditionNames = []string{
+	"IngressStateEndpointsDegraded",
+	"IngressStatePodsDegraded",
+}
+
+func (c *IngressStateController) sync() error {
+	foundConditions := []operatorv1.OperatorCondition{}
+
+	endpoint, err := c.endpointsGetter.Endpoints(c.targetNamespace).Get("oauth-openshift", metav1.GetOptions{})
+	if err != nil {
```

Comment:
can we check for 404 and report it as "no endpoints" ?

Reply:
Done